share-coordinator/src/main/java/org/apache/kafka/coordinator/share/ShareCoordinatorShard.java

```diff
@@ -272,79 +270,89 @@ public CoordinatorResult<WriteShareGroupStateResponseData, CoordinatorRecord> wr
             // since this is the first time we are getting a write request, we should be creating a share snapshot record
-            recordList = Collections.singletonList(ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
-                groupId, topicData.topicId(), partitionData.partition(), ShareGroupOffset.fromRequest(partitionData)
-            ));
+            // the incoming partition data could have overlapping state batches, we must merge them
+            return ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
+                key.groupId(), key.topicId(), partitionData.partition(),
+                new ShareGroupOffset.Builder()
+                    .setSnapshotEpoch(0)
+                    .setStartOffset(partitionData.startOffset())
+                    .setLeaderEpoch(partitionData.leaderEpoch())
+                    .setStateEpoch(partitionData.stateEpoch())
+                    .setStateBatches(getMergedBatches(Collections.emptyList(), partitionData))
+                    .build());
         } else if (snapshotUpdateCount.getOrDefault(key, 0) >= config.shareCoordinatorSnapshotUpdateRecordsPerSnapshot()) {
             int newLeaderEpoch = partitionData.leaderEpoch() == -1 ? shareStateMap.get(key).leaderEpoch() : partitionData.leaderEpoch();
             int newStateEpoch = partitionData.stateEpoch() == -1 ? shareStateMap.get(key).stateEpoch() : partitionData.stateEpoch();
             long newStartOffset = partitionData.startOffset() == -1 ? shareStateMap.get(key).startOffset() : partitionData.startOffset();
 
             // Since the number of update records for this share part key exceeds snapshotUpdateRecordsPerSnapshot,
             // we should be creating a share snapshot record.
-            List<PersisterStateBatch> batchesToAdd = StateBatchUtil.combineStateBatches(
-                shareStateMap.get(key).stateBatches(),
-                partitionData.stateBatches().stream()
-                    .map(PersisterStateBatch::from)
-                    .collect(Collectors.toList()),
-                newStartOffset);
-
-            recordList = Collections.singletonList(ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
-                groupId, topicData.topicId(), partitionData.partition(),
+            return ShareCoordinatorRecordHelpers.newShareSnapshotRecord(
+                key.groupId(), key.topicId(), partitionData.partition(),
                 new ShareGroupOffset.Builder()
+                    .setSnapshotEpoch(shareStateMap.get(key).snapshotEpoch() + 1)   // we must increment snapshot epoch as this is new snapshot
                     .setStartOffset(newStartOffset)
                     .setLeaderEpoch(newLeaderEpoch)
                     .setStateEpoch(newStateEpoch)
-                    .setStateBatches(batchesToAdd)
-                    .build()));
+                    .setStateBatches(getMergedBatches(shareStateMap.get(key).stateBatches(), partitionData, newStartOffset))
```

Comment:
We have a lot of calls to `shareStateMap.get(key)` in this method. Since the key is not changing, we should read this value once at the beginning. 

Also, we should guard against nulls when reading from a map.